versioning/BRM/extentmap.cpp

```diff
@@ -269,19 +270,306 @@ FreeListImpl::FreeListImpl(unsigned key, off_t size, bool readOnly) :
+        constexpr static const size_t minAllowance = 16 * 1024 * 1024;
+        const size_t newShmemSize = std::max(minAllowance, memoryNeeded) + currentShmemSize;
+        grow(newShmemSize);
+        return true;
+    }
+    return false;
+}
+
+InsertUpdateShmemKeyPair ExtentMapIndexImpl::insert(const EMEntry& emEntry, const size_t emIdx)
+{
+    auto dbRoot = emEntry.dbRoot;
+    assert(dbRoot > 0 && dbRoot <= numeric_limits<uint64_t>::max());
+    auto* extentMapIndexPtr = get();
+    assert(extentMapIndexPtr);
+    bool shmemHasGrown = false;
+
+    while (dbRoot >= extentMapIndexPtr->size())
+    {
+        const size_t memNeeded = (extentMapIndexPtr->capacity() + extraUnits_) * dbRootContainerUnitSize_;
+        shmemHasGrown = growIfNeeded(memNeeded);
+        // Need to refresh all refs and iterators b/c the local address range changed.
+        extentMapIndexPtr = get();
+        assert(extentMapIndexPtr);
+        ShmVoidAllocator alloc(fBRMManagedShmMemImpl_.getManagedSegment()->get_segment_manager());
+        OIDIndexContainerT oidIndices(alloc);
+        extentMapIndexPtr->push_back(std::move(oidIndices));
+    }
+    auto& extentMapIndex = *extentMapIndexPtr;
+    return insert2ndLayerWrapper(extentMapIndex[dbRoot], emEntry, emIdx, shmemHasGrown);
+}
+
+InsertUpdateShmemKeyPair ExtentMapIndexImpl::insert2ndLayer(OIDIndexContainerT& oids,
+    const EMEntry& emEntry, const size_t emIdx, const bool aShmemHasGrown)
+{
+    OID_t oid = emEntry.fileID;
+    ShmVoidAllocator alloc(fBRMManagedShmMemImpl_.getManagedSegment()->get_segment_manager());
+
+    PartitionIndexContainerT partitionIndex(alloc);
+    auto iterAndResult = oids.insert({oid, std::move(partitionIndex)});
+
+    if (iterAndResult.second)
+    {
+        PartitionIndexContainerT& partitionsContainer = (*iterAndResult.first).second;
+        return insert3dLayerWrapper(partitionsContainer, emEntry, emIdx, aShmemHasGrown);
+    }
+    else
+        return {false, aShmemHasGrown};
+}
+
+InsertUpdateShmemKeyPair ExtentMapIndexImpl::insert2ndLayerWrapper(OIDIndexContainerT& oids,
+    const EMEntry& emEntry,
+    const size_t emIdx, const bool aShmemHasGrown)
+{
+    OID_t oid = emEntry.fileID;
+    auto oidsIter = oids.find(oid);
+    bool shmemHasGrown = aShmemHasGrown;
+    if (oidsIter == oids.end())
+    {
+        const size_t freeShmem = fBRMManagedShmMemImpl_.getManagedSegment()->get_free_memory();
+        const size_t memNeeded = (oids.size() + extraUnits_) * oidContainerUnitSize_;
+        if (oids.load_factor() >= oids.max_load_factor() ||
+            freeShmem <= freeSpaceThreshold_)
+        {
+            // Need to refresh all refs and iterators b/c the local address range changed.
+            shmemHasGrown = growIfNeeded(memNeeded);
+            auto* extMapIndexPtr = get();
+            assert(extMapIndexPtr);
+            auto& extMapIndex = *extMapIndexPtr;
+            shmemHasGrown = std::max(shmemHasGrown, aShmemHasGrown);
```

Comment:
Why not just `shmemHasGrown || aShmemHasGrown` ?

Reply:
Good point.